include/boost/geometry/algorithms/detail/overlay/split_rings.hpp

```diff
@@ -0,0 +1,433 @@
+        }
+    }
+
+    inline iterator begin() { return m_list.begin(); }
+    inline const_iterator begin() const { return m_list.begin(); }
+
+    inline iterator end() { return m_list.end(); }
+    inline const_iterator end() const { return m_list.end(); }
+
+    inline void swap(ring_as_dcl& other)
+    {
+        m_list.swap(other.m_list);
+    }
+private:
+    List m_list;
+};
+
+
+template <closure_selector Closure>
+class find_duplicate_points
+{
+    static bool const is_closed = Closure == closed;
+
+    struct point_iterator_less
+    {
+        template <typename PointIterator>
+        inline bool operator()(PointIterator it1, PointIterator it2) const
+        {
+            return geometry::less
+                <
+                    typename std::iterator_traits<PointIterator>::value_type
+                >()(*it1, *it2);
+        }
+    };
+
+public:
+    template <typename Ring, typename Iterator>
+    static inline bool apply(Ring& ring, Iterator& pos1, Iterator& pos2)
+    {
+        typedef typename Ring::iterator iterator_type;
+        typedef std::set<iterator_type, point_iterator_less> point_set_type;
+        point_set_type point_set;
+
+        iterator_type last = is_closed ? --boost::end(ring) : boost::end(ring);
+        for (iterator_type it = boost::begin(ring); it != last; ++it)
+        {
+            typename point_set_type::const_iterator pit = point_set.find(it);
+            if (pit != point_set.end())
+            {
+                pos1 = *pit;
+                pos2 = it;
+                return true;
+            }
+            else
+            {
+                point_set.insert(it);
+            }
+        }
+
+        // initialize pos1 and pos2 with something
+        pos1 = boost::begin(ring);
+        pos2 = pos1;
+        return false;
+    }
+};
+
+
+template <overlay_type OverlayType, typename Ring, typename RobustPolicy>
+struct split_ring
+{
+    template <typename RingCollection>
+    static inline void apply(Ring const&, RingCollection&, RobustPolicy const&)
+    {
+    }
+};
+
+// specialization for union
+// TODO: add another specialization for intersection once implemented
+template <typename Ring, typename RobustPolicy>
+class split_ring<overlay_union, Ring, RobustPolicy>
+{
+    typedef turn_info
+        <
+            typename point_type<Ring>::type,
+            typename geometry::segment_ratio_type
+                <
+                    typename point_type<Ring>::type, RobustPolicy
+                >::type
+        > turn_type;
+
+    typedef std::deque<turn_type> turns_container_type;
+
+    struct no_interrupt_policy
+    {
+        static bool const enabled = false;
+        static bool const has_intersections = false;
+
+        template <typename Range>
+        static inline bool apply(Range const&)
+        {
+            return false;
+        }
+    };
+
+    template <typename Turn>
+    static inline
+    typename Turn::turn_operation_type get_correct_op(Turn const& t)
```

Comment:
Maybe the name is not the best one.

Here is what is happening: I am having the turn `t`, which is of m:u/u type, and I want to get back the segment identifier and the fraction corresponding to this turn, and I want to get these two via the proper operation related to this turn. However, I am interested in the segment identifier and the fraction for the segment that the turn point lies in the *interior* of the segment (so the fraction cannot be either 0 or 1), rather than the segment identifier for which the turn point is an endpoint of the segment. So the ***correct*** operation I am interested in is the one for which the fraction is neither 0 nor 1, hence the name of the method.